src/Services/Azure/Authentication/CustomChainedCredential.cs

```diff
@@ -18,22 +20,99 @@ namespace AzureMcp.Services.Azure.Authentication;
+    public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        ChainedTokenCredential chainedCredential = CreateChainedCredential(tenantId);
+        return chainedCredential.GetToken(requestContext, cancellationToken);
+    }
 
-    public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken) =>
-        _chainedCredential.GetToken(requestContext, cancellationToken);
+    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        ChainedTokenCredential chainedCredential = CreateChainedCredential(tenantId);
+        return chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+    }
+
+    private static ChainedTokenCredential CreateChainedCredential(string? tenantId)
+    {
+        string? authRecordJson = Environment.GetEnvironmentVariable("AZURE_MCP_AUTH_AUTHENTICATION_RECORD");
+        AuthenticationRecord? authRecord = null;
+        if (!string.IsNullOrEmpty(authRecordJson))
+        {
+            byte[] bytes = Encoding.UTF8.GetBytes(authRecordJson);
+            using MemoryStream authRecordStream = new MemoryStream(bytes);
+            authRecord = AuthenticationRecord.Deserialize(authRecordStream);
+        }
 
-    public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken) =>
-        _chainedCredential.GetTokenAsync(requestContext, cancellationToken);
+        string? useOnlyBrokerCredential = Environment.GetEnvironmentVariable("AZURE_MCP_USE_ONLY_BROKER_CREDENTIAL");
+        if (useOnlyBrokerCredential == "true")
+        {
+            return new(CreateBrowserCredential(tenantId, authRecord));
+        }
+        else
+        {
+            return new(CreateDefaultCredential(tenantId), CreateBrowserCredential(tenantId, authRecord));
+        }
+    }
+
+    private enum GetAncestorFlags
+    {
+        GetParent = 1,
+        GetRoot = 2,
+        /// <summary>
+        /// Retrieves the owned root window by walking the chain of parent and owner windows returned by GetParent.
+        /// </summary>
+        GetRootOwner = 3
+    }
 
-    private static ChainedTokenCredential CreateChainedCredential(string? tenantId) =>
-        new(CreateDefaultCredential(tenantId), CreateBrowserCredential());
+    /// <summary>
+    /// Retrieves the handle to the ancestor of the specified window.
+    /// </summary>
+    /// <param name="hwnd">A handle to the window whose ancestor is to be retrieved.
+    /// If this parameter is the desktop window, the function returns NULL. </param>
+    /// <param name="flags">The ancestor to be retrieved.</param>
+    /// <returns>The return value is the handle to the ancestor window.</returns>
+    [DllImport("user32.dll", ExactSpelling = true)]
+    static extern IntPtr GetAncestor(IntPtr hwnd, GetAncestorFlags flags);
```

Comment:
You should be using the WindowHandler that is in the tests folder.  You can bring it into product /src folder if need be.